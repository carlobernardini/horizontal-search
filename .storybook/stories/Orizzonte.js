import React from 'react';
import Orizzonte, {
    Choices, Dropdown, FullText, Group, Select, Toggle
} from 'orizzonte';
import { truncate } from 'lodash-es';
import axios from 'axios';
// eslint-disable-next-line import/no-extraneous-dependencies
import MockAdapter from 'axios-mock-adapter';
// eslint-disable-next-line import/no-extraneous-dependencies
import ArrayMove from 'array-move';
// eslint-disable-next-line import/no-extraneous-dependencies
import { storiesOf } from '@storybook/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { withInfo } from '@storybook/addon-info';
// eslint-disable-next-line import/no-extraneous-dependencies
import { withState } from '@dump247/storybook-state';
// eslint-disable-next-line import/no-extraneous-dependencies
import { withKnobs, text, boolean, object, select } from '@storybook/addon-knobs/react';

const stories = storiesOf('Orizzonte', module);
stories.addDecorator(withKnobs);

const mockAPI = new MockAdapter(axios, {
    delayResponse: 750
});
const apiRequest = 'https://orizzonte.io/suggestions';
const remoteOptions = [{
    label: 'United Kingdom',
    value: 'uk'
}, {
    label: 'France',
    value: 'fr'
}, {
    label: 'Germany',
    value: 'de'
}, {
    value: 'Benelux',
    children: [{
        label: 'The Netherlands',
        value: 'nl',
    }, {
        label: 'Belgium',
        value: 'be'
    }, {
        label: 'Luxembourg',
        value: 'lu'
    }]
}, {
    label: 'Spain',
    value: 'es'
}, {
    label: 'Italy',
    value: 'it'
}];

// eslint-disable-next-line react/prop-types
const component = ({ store }) => {
    const { groups, query } = store.state;

    mockAPI.onAny(apiRequest).reply((config) => {
        let filter = null;

        try {
            filter = new RegExp(`(${ JSON.parse(config.data).q })`, 'gi');
        // eslint-disable-next-line no-empty
        } catch (e) {}

        return [200, {
            options: filter
                ? remoteOptions.filter((option) => ((option.label || option.value).match(filter)))
                : remoteOptions
        }];
    });

    return (
        <Orizzonte
            addBtnLabel={ text('Add button label', '') }
            autoHideControls={ boolean('Auto-hide controls (add, clear, save)', true) }
            query={ object('Query', query) }
            collapseGroupOnClickOutside={ boolean('Collapse groups on click outside', true) }
            clearAllLabel={ text('Label for Clear All button', '') }
            clearedQuerySnapshot={{
                language: 'fr',
                calendarPeriod: ['3m', '1y']
            }}
            saveLabel={ text('Label for Save button', '') }
            groupTopLabels={ boolean('Labels on top of groups', true) }
            hideAddOnAllGroupsIncluded={ boolean('Hide add-button when no more groups available', true) }
            dispatchOnFilterChange={ boolean('Dispatch query on filter change', true) }
            orientation={ select('Orientation', ['ltr', 'rtl'], 'ltr') }
            onChange={ (queryObject) => {
                console.log(queryObject);
                store.set({
                    query: queryObject
                });
            }}
            onClearAll={ (clearedQuerySnapshot) => {
                store.set({
                    query: clearedQuerySnapshot
                });
            }}
            onGroupAdd={ (i) => {
                let newGroups = store.state.groups.slice(0);
                newGroups[i].included = true;
                newGroups = ArrayMove(newGroups, i, newGroups.length - 1);

                store.set({
                    groups: newGroups
                });
            }}
            onGroupRemove={ (i) => {
                const newGroups = store.state.groups.slice(0);
                newGroups[i].included = false;

                store.set({
                    groups: newGroups
                });
            }}
            onSave={ (queryObject) => {
                console.log('Save query', queryObject);
            }}
        >
            {
                groups.map((group, i) => {
                    const { filters, ...rest } = group;

                    return (
                        <Group
                            key={ `${ group.name }-${ i }` }
                            { ...rest }
                        >
                            { filters }
                        </Group>
                    );
                })
            }
        </Orizzonte>
    );
};

stories.add('Default', withState({
    query: {
        language: 'fr',
        calendarPeriod: ['3m', '1y'],
        toggle: 'on'
    },
    groups: [{
        included: true,
        label: 'Locale',
        doneBtnLabel: 'OK',
        removeBtnLabel: 'Remove me',
        filters: [
            <Select
                key="language"
                fieldName="language"
                label="Language"
                selectedLabel="%s (Primary)"
                options={ [{
                    value: 'English',
                    children: [{
                        label: 'UK English',
                        value: 'en-gb'
                    }, {
                        label: 'US English',
                        value: 'en-us'
                    }]
                }, {
                    label: 'French',
                    value: 'fr'
                }, {
                    label: 'German',
                    value: 'de',
                    disabled: true
                }, {
                    label: 'Dutch',
                    value: 'nl'
                }] }
            />,
            <Toggle
                key="mytoggle"
                fieldName="toggle"
                option={{
                    label: 'Toggle me on or off',
                    value: 'on'
                }}
                selectedLabel="Toggle is on"
                toggleStateLabel={{
                    on: 'Active',
                    off: 'Inactive'
                }}
            />,
            <Dropdown
                key="country"
                information="Choose one or more countries (optional)"
                fieldName="country"
                label="Country"
                selectedLabel={ (value) => {
                    if (value.length === 1) {
                        return value[0].label;
                    }
                    return `${ value.length } Countries`;
                }}
                options={[{
                    label: 'Austria',
                    value: 'at'
                }, {
                    label: 'Portugal',
                    value: 'pt'
                }, {
                    label: 'Ireland',
                    value: 'ie'
                }]}
                multiple
                filter={{
                    enabled: true,
                    matchPosition: 'start',
                    placeholder: 'Search options...'
                }}
                remote={{
                    endpoint: 'https://orizzonte.io/suggestions',
                    searchParam: 'q',
                    params: {
                        some: 'additional data'
                    },
                    transformer: (response) => (response.options)
                }}
            />,
            <Select
                disabled
                key="secondary-language"
                fieldName="secondLanguage"
                label="Secondary Language"
                options={ [{
                    label: 'English',
                    value: 'en'
                }, {
                    label: 'French',
                    value: 'fr'
                }, {
                    label: 'German',
                    value: 'de'
                }, {
                    label: 'Dutch',
                    value: 'nl'
                }] }
                notSetLabel="None"
            />
        ]
    }, {
        included: true,
        label: 'Sizes',
        filters: [
            <Dropdown
                key="shirt-size"
                fieldName="shirtSize"
                label="Shirt Size"
                selectedLabel={ (value, totalCount) => {
                    if (value.length <= 2) {
                        return `Shirt Size (${ value.map((v) => v.label).join(' & ') })`;
                    }
                    if (value.length === totalCount) {
                        return 'Any shirt size';
                    }
                    return `Shirt Size (${ value.length } selected)`;
                }}
                options={ [{
                    value: 'Small sizes',
                    children: [{
                        label: 'Extra Small',
                        value: 'xs',
                        disabled: true,
                        facetCount: 0
                    }, {
                        label: 'Small',
                        value: 's',
                        facetCount: 129
                    }]
                }, {
                    label: 'Medium',
                    value: 'm',
                    facetCount: 213
                }, {
                    value: 'Large sizes',
                    children: [{
                        label: 'Large',
                        value: 'l',
                        facetCount: 1982
                    }, {
                        label: 'Extra Large',
                        value: 'xl',
                        facetCount: 94
                    }]
                }] }
                notSetLabel="None"
                multiple
                selectAll
                selectAllCount
            />,
            <Choices
                key="waist-size"
                fieldName="waistSize"
                label="Waist Size"
                selectedLabel={ (value) => (`${ value.selectedLabel || value.label } waist size`) }
                noPreferenceLabel="No preference"
                options={ [{
                    label: 'Extra Small (28)',
                    selectedLabel: 'Extra Small',
                    value: 28,
                    facetCount: 345
                }, {
                    label: 'Small (30)',
                    selectedLabel: 'Small',
                    value: 30,
                    facetCount: 12
                }, {
                    label: 'Medium (32)',
                    selectedLabel: 'Medium',
                    value: 32,
                    facetCount: 228
                }, {
                    label: 'Large (34)',
                    selectedLabel: 'Large',
                    value: 34,
                    facetCount: 192,
                    disabled: true
                }, {
                    label: 'Extra Large (36)',
                    selectedLabel: 'Extra Large',
                    value: 36,
                    facetCount: 73,
                    disabled: true
                }] }
            />
        ]
    }, {
        label: 'Keywords',
        hideDone: true,
        listMinWidth: 400,
        filters: [
            <FullText
                key="keywords"
                fieldName="keywords"
                label="Keywords"
                selectedLabel={ (value) => (truncate(value.trim(), {
                    length: 20
                }))}
                placeholder="Enter some keywords, but no numbers..."
                validateInput={ (value) => (
                    !(/[0-9]/g.test(value))
                )}
                autoFocus
                multiline
            />,
            <FullText
                key="disabled"
                disabled
                fieldName="disabledTextField"
                label="Another field"
                placeholder="This one is disabled..."
            />
        ]
    }, {
        included: true,
        label: 'Dates',
        hideClear: true,
        description: 'You can narrow your search to any of the time spans listed below',
        filters: [
            <Choices
                multiple
                key="period"
                fieldName="calendarPeriod"
                selectedLabel={ (value) => {
                    if (value.length === 1) {
                        return value[0].label;
                    }
                    return `Calendar Period (${ value.length } selected)`;
                }}
                options={ [{
                    label: 'Last Month',
                    value: '1m',
                    facetCount: 23,
                    disabled: true
                }, {
                    label: 'Last 2 Months',
                    value: '2m',
                    facetCount: 543,
                    disabled: true
                }, {
                    label: 'Last 3 Months',
                    value: '3m',
                    facetCount: 1228
                }, {
                    label: 'Last 6 Months',
                    value: '6m',
                    facetCount: 1830
                }, {
                    label: 'Last Year',
                    value: '1y',
                    facetCount: 3247
                }] }
            />
        ]
    }, {
        label: 'Price'
    }, {
        label: 'More...'
    }]
})(
    withInfo()(component)
));
