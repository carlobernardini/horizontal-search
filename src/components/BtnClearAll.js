import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { DEFAULT_STR_CLEAR_ALL, DISPLAY_NAME_BTN_CLEAR_ALL } from '../constants';

const BtnClearAll = ({ clearAllLabel, disabled, onClearAll, position, shown }) => (
    <div
        className={ classNames('orizzonte__btn-wrapper', {
            'orizzonte__btn-wrapper--left': position === 'left',
        }) }
    >
        <button
            className={ classNames('orizzonte__btn orizzonte__btn-text', {
                'orizzonte__btn--shown': shown,
                'orizzonte__btn--disabled': disabled
            }) }
            disabled={ disabled }
            type="button"
            onClick={ () => {
                onClearAll();
            }}
        >
            { clearAllLabel && clearAllLabel.length ? clearAllLabel : DEFAULT_STR_CLEAR_ALL }
        </button>
    </div>
);

BtnClearAll.displayName = DISPLAY_NAME_BTN_CLEAR_ALL;

BtnClearAll.propTypes = {
    clearAllLabel: PropTypes.string,
    disabled: PropTypes.bool.isRequired,
    onClearAll: PropTypes.func.isRequired,
    position: PropTypes.string.isRequired,
    shown: PropTypes.bool.isRequired
};

BtnClearAll.defaultProps = {
    clearAllLabel: null
};

export default BtnClearAll;
