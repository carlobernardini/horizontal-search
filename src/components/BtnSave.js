import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const BtnSave = ({ saveLabel, disabled, onSave, position, shown }) => (
    <div
        className={ classNames('orizzonte__btn-wrapper', {
            'orizzonte__btn-wrapper--left': position === 'left',
        }) }
    >
        <button
            className={ classNames('orizzonte__btn orizzonte__btn-text', {
                'orizzonte__btn--shown': shown,
                'orizzonte__btn--disabled': disabled
            }) }
            disabled={ disabled }
            type="button"
            onClick={ () => {
                onSave();
            }}
        >
            { saveLabel && saveLabel.length ? saveLabel : 'Save' }
        </button>
    </div>
);

BtnSave.propTypes = {
    saveLabel: PropTypes.string,
    disabled: PropTypes.bool.isRequired,
    onSave: PropTypes.func.isRequired,
    position: PropTypes.string.isRequired,
    shown: PropTypes.bool.isRequired
};

BtnSave.defaultProps = {
    saveLabel: null
};

export default BtnSave;
