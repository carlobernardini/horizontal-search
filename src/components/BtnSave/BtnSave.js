import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { DEFAULT_STR_SAVE, DISPLAY_NAME_BTN_SAVE } from '../../constants';

const BtnSave = ({ saveLabel, disabled, onSave, position, shown }) => (
    <div
        className={ classNames('orizzonte__btn-wrapper', {
            'orizzonte__btn-wrapper--left': position === 'left',
        }) }
    >
        <button
            className={ classNames('orizzonte__btn orizzonte__btn-text', {
                'orizzonte__btn--shown': shown,
                'orizzonte__btn--disabled': disabled
            }) }
            disabled={ disabled }
            type="button"
            onClick={ () => {
                onSave();
            }}
        >
            { saveLabel && saveLabel.length ? saveLabel : DEFAULT_STR_SAVE }
        </button>
    </div>
);

BtnSave.displayName = DISPLAY_NAME_BTN_SAVE;

BtnSave.propTypes = {
    saveLabel: PropTypes.string,
    disabled: PropTypes.bool.isRequired,
    onSave: PropTypes.func.isRequired,
    position: PropTypes.string.isRequired,
    shown: PropTypes.bool.isRequired
};

BtnSave.defaultProps = {
    saveLabel: null
};

export default BtnSave;
