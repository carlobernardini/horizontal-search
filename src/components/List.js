import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import '../scss/List.scss';

class List extends Component {
    renderDoneBtn() {
        const {
            doneBtn, doneBtnLabel, isFilterGroup, onApply
        } = this.props;

        if (!isFilterGroup || !doneBtn) {
            return null;
        }

        return (
            <li>
                <button
                    type="button"
                    className="orizzonte__list-done"
                    onClick={ onApply }
                >
                    { doneBtnLabel || 'Done' }
                </button>
            </li>
        );
    }

    renderItems() {
        const {
            values, items, isFilterGroup, onUpdate
        } = this.props;

        if (isFilterGroup) {
            return React.Children.map(items, (item, i) => (
                <li
                    className={ classNames('orizzonte__item', {
                        'orizzonte__item--filters': isFilterGroup
                    }) }
                    key={ i }
                >
                    { React.cloneElement(item, {
                        value: ((v, fn) => {
                            if (!(fn in v)) {
                                return null;
                            }
                            return v[fn];
                        })(values, item.props.fieldName),
                        onUpdate: (filterValue) => {
                            const { fieldName } = item.props;
                            onUpdate(fieldName, filterValue);
                        }
                    }) }
                </li>
            ));
        }

        return items.map((item, i) => (
            <li
                className={ classNames('orizzonte__item', {
                    'orizzonte__item--filters': isFilterGroup
                }) }
                key={ i }
            >
                { item }
            </li>
        ));
    }

    render() {
        const { orientation } = this.props;

        return (
            <ul
                className={ classNames('orizzonte__list', {
                    'orizzonte__list--right': orientation === 'right'
                }) }
            >
                { this.renderItems() }
                { this.renderDoneBtn() }
            </ul>
        );
    }
}

List.propTypes = {
    doneBtn: PropTypes.bool,
    doneBtnLabel: PropTypes.string,
    values: PropTypes.object,
    isFilterGroup: PropTypes.bool,
    items: PropTypes.array.isRequired,
    onApply: PropTypes.func,
    onUpdate: PropTypes.func,
    orientation: PropTypes.oneOf([
        'left',
        'right'
    ])
};

List.defaultProps = {
    doneBtn: true,
    doneBtnLabel: null,
    values: {},
    isFilterGroup: false,
    onApply: () => {},
    onUpdate: () => {},
    orientation: 'left'
};

export default List;
