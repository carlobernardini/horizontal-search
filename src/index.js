export { default } from './components/Orizzonte';
export { default as Group } from './components/Group';
export { default as Select } from './components/Select';
