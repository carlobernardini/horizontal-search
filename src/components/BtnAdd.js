import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { DISPLAY_NAME_BTN_ADD } from '../constants';
import List from './List';
import '../scss/Button.scss';

class BtnAdd extends PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            active: false
        };
        this.btnAdd = React.createRef();
        this.toggleButton = this.toggleButton.bind(this);
        document.addEventListener('click', this.onBodyClick.bind(this), false);
    }

    componentWillUnmount() {
        document.removeEventListener('click', this.onBodyClick.bind(this), false);
    }

    onBodyClick(e) {
        if (!this.btnAdd || !this.btnAdd.current || this.btnAdd.current.contains(e.target)) {
            return false;
        }

        this.setState({
            active: false
        });

        return true;
    }

    toggleButton() {
        const { active } = this.state;

        this.setState({
            active: !active
        });
    }

    renderList() {
        const { active } = this.state;
        const { available, onGroupAdd, position, shown } = this.props;

        if (!shown || !active) {
            return null;
        }

        return (
            <List
                orientation={ position === 'right' ? 'left' : 'right' }
            >
                { available.map((filter, i) => (
                    // eslint-disable-next-line jsx-a11y/anchor-is-valid
                    <a
                        key={ `orizzonte-add-${ i }` }
                        href="#"
                        className="orizzonte__item-clickable"
                        onClick={ (e) => {
                            e.preventDefault();
                            onGroupAdd(filter.i, filter.label);
                            this.toggleButton();
                        }}
                    >
                        { filter.label }
                    </a>
                )) }
            </List>
        );
    }

    render() { 
        const {
            available, disabled, label, position, shown
        } = this.props;

        return (
            <div
                className={ classNames('orizzonte__btn-wrapper', {
                    'orizzonte__btn-wrapper--left': position === 'left'
                }) }
                ref={ this.btnAdd }
            >
                <button
                    className={ classNames('orizzonte__btn orizzonte__btn-add', {
                        'orizzonte__btn--shown': shown,
                        'orizzonte__btn--disabled': disabled || !available.length,
                        'orizzonte__btn--labeled': label && label.trim().length
                    }) }
                    disabled={ !available.length }
                    onClick={ this.toggleButton }
                    type="button"
                >
                    { label || String.fromCharCode(160) }
                </button>
                { this.renderList() }
            </div>
        );
    }
}

BtnAdd.displayName = DISPLAY_NAME_BTN_ADD;

BtnAdd.propTypes = {
    available: PropTypes.array,
    disabled: PropTypes.bool,
    label: PropTypes.string,
    onGroupAdd: PropTypes.func,
    position: PropTypes.string,
    shown: PropTypes.bool
};

BtnAdd.defaultProps = {
    available: [],
    disabled: false,
    label: null,
    onGroupAdd: () => {},
    position: null,
    shown: false
};

export default BtnAdd;
