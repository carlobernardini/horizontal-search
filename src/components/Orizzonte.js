import React, { Component } from 'react';
import PropTypes from 'prop-types';
import BtnAdd from './BtnAdd';
import '../scss/Orizzonte.scss';

class Orizzonte extends Component {
    renderAddBtn(position) {
        const { btnAddPosition, children, maxFilters } = this.props;

        if (btnAddPosition !== position) {
            return null;
        }

        if (maxFilters && maxFilters === React.Children.count(children)) {
            return null;
        }

        return (
            <BtnAdd
                position={ btnAddPosition }
            />
        );
    }

    render() {
        const { children, onFilterRemove } = this.props;

        return (
            <div
                className="orizzonte__container"
            >
                { this.renderAddBtn('left') }
                { React.Children.map(children, (child, i) => {
                    if (child.type.name !== 'Filter') {
                        return null;
                    }

                    return React.cloneElement(child, {
                        i,
                        onFilterRemove
                    });
                }) }
                { this.renderAddBtn('right') }
            </div>
        );
    }
}

Orizzonte.propTypes = {
    btnAddPosition: PropTypes.oneOf([
        'left',
        'right'
    ]),
    children: PropTypes.array,
    disabled: PropTypes.bool,
    maxFilters: PropTypes.number,
    onFilterAdd: PropTypes.func,
    onFilterRemove: PropTypes.func
};

Orizzonte.defaultProps = {
    btnAddPosition: 'right',
    children: [],
    disabled: false,
    maxFilters: null,
    onFilterAdd: () => {},
    onFilterRemove: () => {}
};

export default Orizzonte;
