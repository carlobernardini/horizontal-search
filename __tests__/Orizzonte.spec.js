import React from 'react';
import Orizzonte from '../src/components/Orizzonte';
import { Group, Select } from '../src';
import BtnAdd from '../src/components/BtnAdd';
import BtnClearAll from '../src/components/BtnClearAll';
import BtnSave from '../src/components/BtnSave';
import List from '../src/components/List';

describe('<Orizzonte />', () => {
    it('should render an empty filter container', () => {
        const wrapper = shallow(
            <Orizzonte />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a simple filter container with custom className / styles', () => {
        const query = {
            testSelect: 2
        };
        const clearedQuerySnapshot = {
            testSelect: 1
        };
        const onClearAll = jest.fn();
        const onSave = jest.fn();

        const wrapper = shallow(
            <Orizzonte
                className="custom__class"
                style={{
                    border: '3px solid #000'
                }}
                query={ query }
                onClearAll={ onClearAll }
                onSave={ onSave }
                clearedQuerySnapshot={ clearedQuerySnapshot }
            >
                <Group
                    i={ 1 }
                    label="Test group"
                    groupTopLabels
                    included
                >
                    <Select
                        fieldName="testSelect"
                        label="Test select"
                        options={ [{
                            label: 'Test value 1',
                            value: 1
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }] }
                    />
                </Group>
            </Orizzonte>
        );

        expect(wrapper).toMatchSnapshot();

        wrapper.find(Group).first().prop('onGroupToggle')(1);
        expect(wrapper.state().activeGroup).toBe(1);

        wrapper.find('.orizzonte__container').simulate('mouseover');
        expect(wrapper.state().showControls).toBe(true);

        wrapper.find(BtnClearAll).prop('onClearAll')();
        expect(onClearAll).toHaveBeenCalledWith(clearedQuerySnapshot);

        wrapper.find(BtnSave).prop('onSave')();
        expect(onSave).toHaveBeenCalledWith(query);
    });

    it('should unset an active group when clicking outside of the main component', () => {
        const map = {}

        const onGroupRemove = jest.fn();

        document.addEventListener = jest.fn((event, callback) => {
            map[event] = callback;
        });

        const wrapper = mount(
            <Orizzonte
                className="custom__class"
                style={{
                    border: '3px solid #000'
                }}
                collapseGroupOnClickOutside
                maxGroups={ 1 }
                onGroupRemove={ onGroupRemove }
                autoHideTimeout={ 300 }
            >
                <Group
                    i={ 1 }
                    label="Test group"
                    groupTopLabels
                    included
                >
                    <Select
                        fieldName="testSelect"
                        label="Test select"
                        options={ [{
                            label: 'Test value 1',
                            value: 1
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }] }
                    />
                </Group>
            </Orizzonte>
        );

        const instance = wrapper.instance();

        expect(instance.renderAddBtn()).toBeNull();

        map.click({
            target: null
        });

        expect(wrapper.state().activeGroup).toBeNull();

        wrapper.setState({
            activeGroup: null
        });

        expect(instance.toggleGroup(false)).toBe(false);

        wrapper.find(Group).first().prop('onGroupRemove')(0)
        expect(onGroupRemove).toHaveBeenCalledWith(0);

        jest.useFakeTimers();
        wrapper.setState({
            showControls: true
        });
        expect(wrapper.instance().toggleControls(false)).toBe(true);
        jest.runAllTimers();
        expect(wrapper.state('showControls')).toBe(false);
    });

    it('should test top-level element event handlers', () => {
        const InvalidNode = () => (
            <div />
        );

        const wrapper = shallow(
            <Orizzonte
                hideOnAllGroupsIncluded
            >
                <InvalidNode />
            </Orizzonte>
        );

        const instance = wrapper.instance();
        jest.spyOn(instance, 'toggleControls');

        wrapper.simulate('focus');
        expect(instance.toggleControls).toHaveBeenCalledWith(true);

        wrapper.simulate('blur', {
            target: {
                contains: jest.fn()
            }
        });
        expect(instance.toggleControls).toHaveBeenCalledWith(true);

        wrapper.simulate('mouseover');
        expect(instance.toggleControls).toHaveBeenCalledWith(true);

        wrapper.simulate('mouseout');
        expect(instance.toggleControls).toHaveBeenCalledWith(false);

        expect(wrapper).toMatchSnapshot();
    });

    it('should test adding a new group', () => {
        const onGroupAdd = jest.fn();

        const wrapper = shallow(
            <Orizzonte
                onGroupAdd={ onGroupAdd }
            >
                <Group
                    i={ 1 }
                    label="Test group"
                    groupTopLabels
                >
                    <Select
                        fieldName="testSelect"
                        label="Test select"
                        options={ [{
                            label: 'Test value 1',
                            value: 1
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }] }
                    />
                </Group>
            </Orizzonte>
        );

        const instance = wrapper.instance();
        jest.spyOn(instance, 'toggleGroup');

        wrapper.find(BtnAdd).prop('onGroupAdd')(1);
        expect(onGroupAdd).toHaveBeenCalledWith(1);
        expect(instance.toggleGroup).toHaveBeenCalledWith(0);

        jest.clearAllMocks()

        wrapper.setProps({
            autoExpandOnGroupAdd: false
        });

        wrapper.update();

        wrapper.find(BtnAdd).prop('onGroupAdd')(1);
        expect(instance.toggleGroup).not.toHaveBeenCalled();
    });

    it('should render toggle icons and a custom label for the add button', () => {
        const wrapper = shallow(
            <Orizzonte
                addBtnLabel="Add more groups"
                groupToggleIcon={
                    <span>â–¼</span>
                }
            >
                <Group
                    i={ 1 }
                    label="Test group"
                    groupTopLabels
                    included
                >
                    <Select
                        fieldName="testSelect"
                        label="Test select"
                        options={ [{
                            label: 'Test value 1',
                            value: 1
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }] }
                    />
                </Group>
            </Orizzonte>
        );

        expect(wrapper).toMatchSnapshot();
    });
});
