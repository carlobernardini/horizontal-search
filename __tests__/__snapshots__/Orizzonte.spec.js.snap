// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Orizzonte /> should render a simple filter container with custom className / styles 1`] = `
<div
  className="orizzonte__container orizzonte__clearfix custom__class"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "border": "3px solid #000",
    }
  }
>
  <OrizzonteGroup
    active={false}
    className={null}
    collapseGroupOnClickOutside={false}
    description={null}
    dispatchOnFilterChange={false}
    doneBtnLabel={null}
    groupTopLabels={false}
    hideClear={false}
    hideDone={false}
    hideRemove={false}
    i={0}
    included={true}
    initialState={
      Object {
        "testSelect": 1,
      }
    }
    key=".0"
    label="Test group"
    listMinWidth={null}
    mutuallyExclusiveFilters={false}
    onGroupRemove={[Function]}
    onGroupToggle={[Function]}
    onUpdate={[Function]}
    orientation="left"
    queryPart={
      Object {
        "testSelect": 2,
      }
    }
    removeBtnLabel={null}
    style={Object {}}
  >
    <OrizzonteSelect
      disabled={false}
      fieldName="testSelect"
      information={null}
      label="Test select"
      notSetLabel={false}
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
        ]
      }
      value=""
    />
  </OrizzonteGroup>
  <OrizzonteBtnAdd
    available={Array []}
    disabled={false}
    label={null}
    onGroupAdd={[Function]}
    position="right"
    shown={true}
  />
  <OrizzonteBtnClearAll
    clearAllLabel={null}
    disabled={false}
    onClearAll={[Function]}
    position="right"
    shown={true}
  />
  <OrizzonteBtnSave
    disabled={false}
    onSave={[Function]}
    position="right"
    saveLabel={null}
    shown={true}
  />
</div>
`;

exports[`<Orizzonte /> should render an empty filter container 1`] = `
<div
  className="orizzonte__container orizzonte__clearfix"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={null}
>
  <OrizzonteBtnAdd
    available={Array []}
    disabled={false}
    label={null}
    onGroupAdd={[Function]}
    position="right"
    shown={true}
  />
</div>
`;
